src/context/ConsentContext/ConsentContextFactory.tsx

```diff
@@ -0,0 +1,204 @@
+    isConsentInitialized: false,
+    isConsentGranted: false,
+    redirectToConsentService: () => {},
+  };
+}
+
+/**
+ * Creates a ConsentProvider for given context
+ *
+ * @param context
+ * @returns
+ */
+function getConsentProvider(
+  dataSource: ConsentDataSource,
+  ConsentContext: Context<ConsentContextState<ConsentDataSource>>
+) {
+  return function ConsentProvider({ children }: React.PropsWithChildren) {
+    const toast = useToast();
+
+    //
+    // Reactive flags and states
+    //
+    const [consentSituation, setConsentSituation] = useState<ConsentSituation>({
+      consentStatus: '',
+    });
+    const [isConsentInitialized, setIsConsentInitialized] =
+      useState<boolean>(false);
+    const [isConsentGranted, setIsConsentGranted] = useState<boolean>(false);
+
+    //
+    // Actions
+    //
+    const initializeConsentSituation = useCallback(async () => {
+      if (!isConsentInitialized) {
+        try {
+          await fetchCurrentConsentSituation();
+          setIsConsentInitialized(true);
+        } catch (error) {
+          console.error(error);
+          toast({
+            title: 'Error.',
+            description: "Couldn't initialize consent situation.",
+            status: 'error',
+            duration: 5000,
+            isClosable: true,
+          });
+        }
+      }
+    }, [toast, isConsentInitialized]);
+
+    useEffect(() => {
+      initializeConsentSituation();
+    }, [initializeConsentSituation]);
+
+    //
+    // External service calls
+    //
+    async function fetchCurrentConsentSituation() {
+      const consentStoreKey = `consent-${getEnumKeyFromValue(
+        ConsentDataSource,
+        dataSource
+      )}`;
+
+      const storedConsentToken = sessionStorage.getItem(consentStoreKey);
+      const responseSituation = await api.consent.checkConsent(
+        dataSource,
+        storedConsentToken
+      );
+
+      if (responseSituation.consentToken) {
+        sessionStorage.setItem(consentStoreKey, responseSituation.consentToken);
+      } else if (storedConsentToken) {
+        sessionStorage.removeItem(consentStoreKey);
+      }
+
+      setConsentSituation(responseSituation);
+      setIsConsentGranted(responseSituation.consentStatus === 'consentGranted');
+    }
+
+    function redirectToConsentService(): void {
+      api.consent.directToConsentService(consentSituation);
+    }
+
+    //
+    // URL redirect response handlers
+    //
+    const { search } = useLocation();
+    const queryParams = useMemo(() => new URLSearchParams(search), [search]);
+    const clearQueryParamsFromUrl = queryParams.get('clear');
+
+    /**
+     * On a query param: 'clear', remove query params from the current URL.
+     * @TODO: maybe refactor to some general hook handler
+     */
+    useEffect(() => {
+      if (clearQueryParamsFromUrl === 'true') {
+        window.location.href =
+          window.location.origin + window.location.pathname;
+      }
+    }, [clearQueryParamsFromUrl]);
```

Comment:
Muuttaisin tänän vielä jotenkin näin:

`
const navigation = useNavigation();

 useEffect(() => {
      if (clearQueryParamsFromUrl === 'true') {
        navigate(window.location.pathname);
      }
    }, [clearQueryParamsFromUrl, navigate]);
`

Näin ollen kun tullaan takaisin sivulle, niin ei tapahdu sivun refreshiä ohjauksen yhteydessä.

Tuohon liittyen vielä, kun tuo 'checkConsent' ajellaan, niin riippuen siitä ollaanko vacancies/profiili sivulla, niin asetetaan redirect urli. Eli jos tuo checkConsent ajettiin /vacancies pathissa, ja menen profiilisivulle antamaan consentin -> ohjaa redirectin yhteydessä vacancies sivulle.